include/git2/mailmap.h

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (C) the libgit2 contributors. All rights reserved.
+ *
+ * This file is part of libgit2, distributed under the GNU GPL v2 with
+ * a Linking Exception. For full terms see the included COPYING file.
+ */
+#ifndef INCLUDE_git_mailmap_h__
+#define INCLUDE_git_mailmap_h__
+
+#include "common.h"
+#include "types.h"
+#include "buffer.h"
+
+/**
+ * @file git2/mailmap.h
+ * @brief Mailmap parsing routines
+ * @defgroup git_mailmap Git mailmap routines
+ * @ingroup Git
+ * @{
+ */
+GIT_BEGIN_DECL
+
+/**
+ * Allocate a new mailmap object.
+ *
+ * This object is empty, so you'll have to add a mailmap file before you can do
+ * anything with it. The mailmap must be freed with 'git_mailmap_free'.
+ *
+ * @param out pointer to store the new mailmap
+ * @return 0 on success, or an error code
+ */
+GIT_EXTERN(int) git_mailmap_new(git_mailmap **out);
+
+/**
+ * Free the mailmap and its associated memory.
+ *
+ * @param mm the mailmap to free
+ */
+GIT_EXTERN(void) git_mailmap_free(git_mailmap *mm);
+
+/**
+ * Add a single entry to the given mailmap object. If the entry already exists,
+ * it will be replaced with the new entry.
+ *
+ * @param mm mailmap to add the entry to
+ * @param real_name the real name to use, or NULL
+ * @param real_email the real email to use, or NULL
+ * @param replace_name the name to replace, or NULL
+ * @param replace_email the email to replace
+ * @return 0 on success, or an error code
+ */
+GIT_EXTERN(int) git_mailmap_add_entry(
+	git_mailmap *mm, const char *real_name, const char *real_email,
+	const char *replace_name, const char *replace_email);
+
+/**
+ * Parse mailmap entries from a buffer.
+ *
+ * @param mm mailmap to add the entries to
+ * @param buf the buffer to read the mailmap file from
+ * @param len the length of the input buffer [optional: 0 defaults to 'strlen']
+ * @return 0 on success, or an error code
+ */
+GIT_EXTERN(int) git_mailmap_add_buffer(
```

Comment:
I wonder how much utility there is in making this a public method?  It seems like most uses would prefer the `git_mailmap_from_buffer`.  I'd be perfectly happy seeing this be an internal API.

If it does stay public (which is fine, if you have a use for it or think others will) then I do think that the name does not capture exactly what it does (at least when viewed through the lens of the names of our current APIs - `add_buffer` sounds like it will add a buffer, not add _from_ a buffer).

I would propose `git_mailmap_parse`?  Also, the same notes about the `0` len apply as with `git_mailmap_from_buffer`.

Reply:
Sure, I'm fine with making this internal. Not super attached to any of the public APIs other than the from_repo one.